gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/finalization/DetachStrategy.java

```diff
@@ -0,0 +1,156 @@
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.util.detached.DetachedFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * A strategy that manages the properties that will be in the result.
+ *
+ * @author Valentyn Kahamlyk
+ */
+public final class DetachStrategy extends AbstractTraversalStrategy<TraversalStrategy.FinalizationStrategy> implements TraversalStrategy.FinalizationStrategy {
+
+    private DetachOptions detachOptions;
+
+    // todo: is public constructor ok or better make private?
+    public DetachStrategy(final DetachMode detachMode, final String[] properties) {
+        detachOptions = DetachOptions.build().detachMode(detachMode).properties(properties).create();
+    }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        if (traversal.getParent() == EmptyStep.instance()) {
+            final Optional<ProfileSideEffectStep> profileStep = TraversalHelper.getFirstStepOfAssignableClass(ProfileSideEffectStep.class, traversal);
+            final int index = profileStep.map(step -> traversal.getSteps().indexOf(step))
+                    .orElseGet(() -> traversal.getSteps().size());
+            traversal.addStep(index, new DetachElementStep<>(traversal, detachOptions));
+        }
+    }
+
+    public static DetachStrategy create(final Configuration configuration) {
+        return new DetachStrategy(DetachMode.valueOf(configuration.getString(ID_MODE)), configuration.getStringArray(ID_PROPERTIES));
+    }
+
+    public static final String ID_MODE = "mode";
+    public static final String ID_PROPERTIES = "properties";
+
+    @Override
+    public Configuration getConfiguration() {
+        final Map<String, Object> map = new HashMap<>();
+        map.put(STRATEGY, DetachStrategy.class.getCanonicalName());
+        map.put(ID_MODE, this.detachOptions.detachMode);
+        map.put(ID_PROPERTIES, this.detachOptions.properties);
+        return new MapConfiguration(map);
+    }
+
+    public enum DetachMode {
+
+        ALL,
+        CUSTOM,
+        NONE;
+    }
+
+    public static class DetachOptions {
+        private DetachMode detachMode;
+        private String[] properties;
+
+        private DetachOptions(final DetachOptions.Builder builder) {
+            this.detachMode = builder.detachMode;
+            this.properties = builder.properties;
+        }
+
+        public static DetachOptions.Builder build() {
+            return new DetachOptions.Builder();
+        }
+
+        public DetachMode getDetachMode() {
+            return this.detachMode;
+        }
+
+        public String[] getProperties() {
+            return this.properties;
+        }
+
+        public static final class Builder {
+
+            private DetachMode detachMode = DetachMode.NONE;
+            private String[] properties;
```

Comment:
No default value may lead to a potential null pointer exception.

Reply:
null is expected default. Is it better to explicitly assign null?